From b52b58a18e1901f87f7f623a88b21b0ae5809816 Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Sat, 14 Jun 2014 00:05:01 +1000
Subject: [PATCH 11/11] Fix strides size calculation for YV12/NV12 frames when
 the dimensions are odd.

Make sure to get the very last chroma component for each row.
---
 mythtv/libs/libmythtv/mythframe.h      |   10 ++++++----
 mythtv/libs/libmythtv/videobuffers.cpp |    4 ++--
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/mythtv/libs/libmythtv/mythframe.h b/mythtv/libs/libmythtv/mythframe.h
index bdf8f63..efe8f6f 100644
--- a/mythtv/libs/libmythtv/mythframe.h
+++ b/mythtv/libs/libmythtv/mythframe.h
@@ -131,7 +131,7 @@ static inline void init(VideoFrame *vf, VideoFrameType _codec,
         if (FMT_YV12 == _codec || FMT_YUV422P == _codec)
         {
             vf->pitches[0] = width_aligned;
-            vf->pitches[1] = vf->pitches[2] = width_aligned >> 1;
+            vf->pitches[1] = vf->pitches[2] = (width_aligned+1) >> 1;
         }
         else if (FMT_NV12 == _codec)
         {
@@ -156,13 +156,15 @@ static inline void init(VideoFrame *vf, VideoFrameType _codec,
         {
             vf->offsets[0] = 0;
             vf->offsets[1] = width_aligned * _height;
-            vf->offsets[2] = vf->offsets[1] + (vf->offsets[1] >> 2);
+            vf->offsets[2] =
+                vf->offsets[1] + ((width_aligned+1) >> 1) * ((_height+1) >> 1);
         }
         else if (FMT_YUV422P == _codec)
         {
             vf->offsets[0] = 0;
             vf->offsets[1] = width_aligned * _height;
-            vf->offsets[2] = vf->offsets[1] + (vf->offsets[1] >> 1);
+            vf->offsets[2] =
+                vf->offsets[1] + ((width_aligned+1) >> 1) * _height;
         }
         else if (FMT_NV12 == _codec)
         {
@@ -182,7 +184,7 @@ static inline void clear(VideoFrame *vf)
     if (!vf)
         return;
 
-    int uv_height = vf->height >> 1;
+    int uv_height = (vf->height+1) >> 1;
     if (FMT_YV12 == vf->codec)
     {
         memset(vf->buf + vf->offsets[0],   0, vf->pitches[0] * vf->height);
diff --git a/mythtv/libs/libmythtv/videobuffers.cpp b/mythtv/libs/libmythtv/videobuffers.cpp
index aca1ed7..a939db1 100644
--- a/mythtv/libs/libmythtv/videobuffers.cpp
+++ b/mythtv/libs/libmythtv/videobuffers.cpp
@@ -45,7 +45,7 @@ YUVInfo::YUVInfo(uint w, uint h, uint sz, const int *p, const int *o,
     else
     {
         pitches[0] = width_aligned;
-        pitches[1] = pitches[2] = width_aligned >> 1;
+        pitches[1] = pitches[2] = (width_aligned+1) >> 1;
     }
 
     if (o)
@@ -56,7 +56,7 @@ YUVInfo::YUVInfo(uint w, uint h, uint sz, const int *p, const int *o,
     {
         offsets[0] = 0;
         offsets[1] = width_aligned * height;
-        offsets[2] = offsets[1] + (offsets[1] >> 2);
+        offsets[2] = offsets[1] + ((width_aligned+1) >> 1) * ((height+1) >> 1);
     }
 }
 
-- 
1.7.10.2

